### PR TITLE
feat(core): rework client modules lifecycles, officially make API public

### LINKED ISSUE #3399: Document client lifecycle events
## 📚 Documentation

We have client lifecycle events that allow to hook into the navigation, but it's not documented.

https://github.com/facebook/docusaurus/pull/1591

Example usage in the google analytics plugin:

image

### PR DESCRIPTION
**Pull request by Josh-Cena**
## Motivation

Fix #3399. Client modules suffer from the issue of firing too early—before React even renders. In this case, people who use client modules to modify DOM on every page, or simply reading from it (like the route announcer in #7074) will not be able to do so without using a timeout hack.

This PR also addresses a long-standing issue of hash links not working in development. The problem is that the scroll-to-anchor happens in `shouldComponentUpdate`, so if there's no pre-rendered HTML, there will be no DOM element to read from. This is solved through deferring it to after the initial render as well.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I added a dogfooding client module. When switching between routes, note how the lifecycle now logs the new title instead of the old one: 

image

### CODE REVIEW

File: packages/docusaurus/src/client/PendingNavigation.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/PendingNavigation.tsx:97]: Note that I don't want to have a subcomponent just to dispatch this lifecycle.

1. The `onRouteUpdateDelayed` has to be dispatched in the `PendingNavigation` (because it's fired within a promise callback), so let's just dispatch everything here to avoid scattered code.
2. The `render()` of this component directly returns a `Route`, while having an extra layer makes the architecture not so elegant anymore...

Comment (slorber) [reply packages/docusaurus/src/client/PendingNavigation.tsx:97]: 1) there's something important to consider: do we want to fire the event sync or async? useEffect or useLayoutEffect? (componentDidMount is closer to useLayoutEffect)

I'd rather immediately use a hook today because when we'll refactor this component later we won't risk to change this async vs sync behavior (unless you find a way to test this 😉 ). People may rely on this and I'd rather be explicit about it (even in our docs)

https://reactjs.org/docs/hooks-effect.html

image

IMHO by default this event will be quite used by analytics and reporting. For performance reasons I'd rather it to be async, just after the page transition completed, instead of blocking it. Hence using `useEffect` which is not possible to achieve with a class.

There might be value in having a sync lifecycle that blocks the transition but let's see if users come with a need for this first.

Also, something to consider: what happens when the lifecycle callback throws? I believe the behavior wouldn't be the same in both cases. 

Do we want to fail-fast or fail-safe if an analytics plugin has a bug?

2) I believe this should work

```tsx
      <Route
        location={normalizeLocation(location)}
        render={() => (
          <>
            {children}
            <ClientLifecyclesDispatcher />
          </>
        )}
      />
```

File: website/_dogfooding/clientModuleExample.ts
Comment (slorber) [inline website/_dogfooding/clientModuleExample.ts:19]: it's not before dom repaints imho, it's more before React even starts to render

before dom repaints is useLayoutEffect (and we might not need this immediately)

File: packages/docusaurus/src/client/PendingNavigation.tsx
Comment (slorber) [inline packages/docusaurus/src/client/PendingNavigation.tsx:66]: not 100% sure about this comparison function (as commented in other issues), but we can give it a try for now

maybe we should factorize this, as it's in 2 places and could diverge

File: website/_dogfooding/clientModuleExample.ts
Comment (slorber) [inline website/_dogfooding/clientModuleExample.ts:25]: what about adding this but only for local/deploy previews? we could merge without having logs in prod

+ may be useful to also add onRouteUpdateDelayed too and factorize a bit?

**Review by slorber — CHANGES_REQUESTED** [review]
Looks like a useful change to do that 👍 

See comments: I'd rather really use `useEffect` today as most usage is analytics IMHO, should be non-blocking (perf) and should be fail-safe. Maybe the lifecycle should somehow respect React terminology? Like `onRouteUpdateEffect` + eventually `onRouteUpdateLayoutEffect`?

Also worth migrating our GA plugins to this new lifecycle?


Docs? or do we dogfood this a bit before writing it? 
(considering it impacts the plugin authors we should rather get it right the fist time)

File: website/_dogfooding/clientModuleExample.ts
Comment (Josh-Cena) [reply website/_dogfooding/clientModuleExample.ts:25]: Yes, we would definitely revert this before merging.

> and may be useful to also add `onRouteUpdateDelayed` too and factorize a bit?

That lifecycle's a bit hard to emulate, but we can see what can be meaningfully done in that case (not retrieving DOM)

**Comment by Josh-Cena** [comment]
> `useEffect` or `useLayoutEffect`?

If we are talking about firing analytics, then it doesn't really matter, because the effect doesn't affect UI anyways, it just leaves.

If we are talking about manipulating DOM (for example, someone has presented a weird use-case to me where they have an external script that needs to manipulate a particular DOM element on every page transition), then it certainly needs to be synchronous. I can't think of a compelling case where it unacceptably blocks render.

File: website/_dogfooding/clientModuleExample.ts
Comment (Josh-Cena) [reply website/_dogfooding/clientModuleExample.ts:19]: React commit -> DOM repaint so before React rerender is indeed before DOM repaint :D I see where you are coming from

File: packages/docusaurus/src/client/PendingNavigation.tsx
Comment (Josh-Cena) [reply packages/docusaurus/src/client/PendingNavigation.tsx:97]: > Do we want to fail-fast or fail-safe if an analytics plugin has a bug?

It would be up to the user to catch their error. We just fire the lifecycle.

> I believe this should work

Looks a bit too cluttered IMO, that's what I'm trying to avoid... Could this work?

```js
<ClientLifecyclesDispatcher>
  <Route />
< ClientLifecyclesDispatcher />
```

Also, we always have to put `onRouteUpdateDelayed` within the promise callback, which means the lifecycle dispatching will now be scattered within two components.

Comment (slorber) [reply packages/docusaurus/src/client/PendingNavigation.tsx:97]: > It would be up to the user to catch their error. We just fire the lifecycle.

If it's up to the user (and plugin authors) then we must document that appropriately: if their lifecycle take a long time or if it can throw they must consider catching errors and execution the code asynchronously

I believe it's better if we do this by default instead of letting the user have to think about this. 

> Could this work?

Not 100% but after taking a look it should also work yes. There's probably no value in using children here:

```tsx
<>
  <ClientLifecyclesDispatcher/>
  <Route />
</>
```

> which means the lifecycle dispatching will now be scattered within two components.

That doesn't seem like a big deal considering this is temporary (we'll refactor this later) and the lifecycles behave exactly how we want them to

File: website/_dogfooding/clientModuleExample.ts
Comment (slorber) [reply website/_dogfooding/clientModuleExample.ts:19]: I understand it as "just before dom repaint, after React renders" (ie useLayoutEffect)

image

File: packages/docusaurus/src/client/PendingNavigation.tsx
Comment (Josh-Cena) [reply packages/docusaurus/src/client/PendingNavigation.tsx:97]: > if it can throw they must consider catching errors and execution the code asynchronously

The client-side APIs are usually async already. Is there anything that's not callback-/promise-based but render-blocking? Can't think of a compelling use-case.

**Comment by slorber** [comment]
> If we are talking about firing analytics, then it doesn't really matter, because the effect doesn't affect UI anyways, it just leaves.

Actually it can happen UI:

- the analytics SDK could fail, leading to the page transition never completing
- the analytics SDK could be slow, serialize synchronously a lot of JSON or whatever, leading to a delayed page transition UX

That's why I think `useEffect` is a better default. This is also why the React team created `useEffect` and `useLayoutEffect` and we use `useEffect` in most cases by default. They could have given us `useLayoutEffect` only and told us to use `setTimeout` but they didn't for good reasons: help us do the right thing by default.

> someone has presented a weird use-case to me where they have an external script that needs to manipulate a particular DOM element on every page transition

Then we need to understand better this use-case

**Comment by Josh-Cena** [comment]
> Then we need to understand better this use-case

I remember it was attaching event listeners.

```js
import ExecutionEnvironment from '@docusaurus/ExecutionEnvironment';

export default (function () {
  if (!ExecutionEnvironment.canUseDOM) {
    return null;
  }

  return {
    onRouteUpdate({location}) {
      if (location.pathname.startsWith('/acs/')) {
        const labels = document.getElementsByTagName("label");
        for (let a = 0; a < labels.length; a++)
          labels[a].addEventListener("click", (e) => e.target.classList.add("worngans"));
        document.querySelector('input,textarea').attr('autocomplete', 'off');
      }
    },
  };
})();
```

**Comment by Josh-Cena** [comment]
@slorber After some more careful investigation, I think the current code is good enough. Anything manipulating the DOM in `onRouteDidUpdate` needs to be fired synchronously anyways, so putting it in `componentDidUpdate` makes more sense. This lifecycle is needed in the first place because some users use it to get hold of the page's DOM. We can't do that with `useEffect` because we can't cleanup. Analytics and other things that are purely side-effects do not block render, and therefore, whether promise- or callback-based, do not matter if we put them in an async or sync context. Could you give a concrete example where the current sync handling would be problematic?

We are implementing lifecycles anyways, not hooks. Lifecycles are always meant to be called "at a certain point", just like React lifecycles. I think the semantics are quite clear, we just need documentation about error handling, etc.

File: packages/docusaurus/src/client/PendingNavigation.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/PendingNavigation.tsx:86]: Moving this scrolling to `componentDidUpdate` fixes the long-standing issue that hash links don't work cross-page in development. Now, when calling `scrollIntoView`, the anchored DOM always exists even when it's the first render.

**Comment by Josh-Cena** [comment]
@slorber I implemented what you had in mind in https://github.com/facebook/docusaurus/pull/6732/commits/a4e16b9307345f9b6fb556fc13e33eca6e54cd80. Tell me what you think and if that's the current direction we should go in.

I realized because `useLayoutEffect` still fires after React renders, even if we put `onRouteUpdate` in `useLayoutEffect`, the lifecycle is able to get the latest DOM on the new page. Therefore I think it's not necessary to make another lifecycle. If we are to continue supporting this lifecycle firing before React re-render, we have to put it in `shouldComponentUpdate`.

File: packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx:58]: I'm not sure whether putting this in `useLayoutEffect` or `useEffect` is better semantic. Because client modules don't have access to the React context and they contain purely imperative logic (again, I don't think analytics can possibly block rendering anyways), putting them in either places make sense.

File: packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx:55]: Moving this logic from `shouldComponentUpdate` to here fixes the long-standing issue of hash links not working cross-page in development, because now when `scrollIntoView` into view, the anchor DOM always exists.

File: packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx:62]: This has to be fired from the timeout callback, so just exposed an imperative handle.

File: packages/docusaurus/src/client/PendingNavigation.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/PendingNavigation.tsx:69]: This should now look much cleaner: no more DOM access in `shouldComponentUpdate` 🎉

File: packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx
Comment (Josh-Cena) [reply packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx:58]: Note that I **intentionally** left out the `pathname === pathname` check. I think this should be handled on the downstream because who knows? Some client modules may find this useful.

**Review by slorber — COMMENTED** [review]
LGTM 👍

File: packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx
Comment (Josh-Cena) [reply packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx:59]: I looked into it, but not sure if it's even possible... The only alternative I can think of is `React.memo`, but that doesn't have access to the component state in the `propHasChanged` callback

Comment (Josh-Cena) [reply packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx:59]: I looked into it, but not sure if it's even possible... The only alternative I can think of is `React.memo`, but that doesn't have access to the component state in the `propHasChanged` callback

**Comment by Josh-Cena** [comment]
@slorber There are a few more ideas on my mind. Not necessarily have to be in this PR, but just to ask if you are interested:

1. Can we make nprogress a client module, and move it to theme-classic? This way, we can make the nprogress more configurable than it currently is (because the core is not configurable), and it can also utilize the site's theme color to provide consistent styles.
2. If we are to make nprocess a client module, we have to have a "clean up" function for `onRouteUpdateDelayed`. `nprogress.start()` is called along with `onRouteUpdateDelayed`, but `nprogress.done()` is called when the route finished loading, and we don't have a lifecycle from it yet. What about returning a clean up function like effects do?

```js
export function onRouteUpdateDelayed({ location }) {
  nprogress.start();
  return () => nprogress.done();
}
```

**Comment by slorber** [comment]
We could even extract this as a plugin and provide configuration options.

Gatsby does so: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-nprogress/src/gatsby-browser.js

Not sure the cleanup callback is really needed, as it somehow becomes duplicated with `onRouteUpdate` which is supposed to always fire after `onRouteUpdateDelayed`. Cf Gatsby implementation, they don't need it either.


---

Now, we could eventually rethink the whole thing and only have a single lifecycle to handle everything, but that may not be very intuitive for plugin authors, and also a more complex breaking change:

```js
export function onRouteUpdate({ location }) {

  console.log("route transition triggered")

  const timeout = setTimeout(() => {
    console.log("route transition delayed")
    nprogress.start();
  },200)

  return () => {
    console.log("route transition completed")
    clearTimeout(timeout);
    nprogress.done();
  }

}
```

**Comment by Josh-Cena** [comment]
Ah, that makes sense as well. Looking back, the entire timeout is only used for the nprogress.

And yes, VuePress also extracts nprogress to a separate plugin, that's why I'm asking...

Will merge it shortly after

**Comment by slorber** [comment]
👍 

Actually thinking about it, that wouldn't even be a breaking change considering onRouteUpdate already fires too early today 😄 

maybe it's the best time to decide on whether we want 2 lifecycles or just 1 more powerful with a cleanup function?

**Comment by Josh-Cena** [comment]
Let's move nprogress to theme-classic, shall we? This way we can make the core even more lightweight.

**Comment by slorber** [comment]
> Let's move nprogress to theme-classic, shall we? This way we can make the core even more lightweight.

ok for that 👍

**Comment by slorber** [comment]
We agreed on Discord to try this API:

```js
// fire early
export const onRouteWillUpdate = () => {
  
  const nprogressTimeout = setTimeout(() => {
    NProgress.start()
  }, 200); // timeout can be configured
  
  // cleanup function is fired when the update is ready to be done (ie next page bundles have been fetched etc...)
  return () => clearTimeout(nprogressTimeout);
}

// fire after React renders in useLayoutEffect
export const onRouteUpdate = () => {
  NProgress.done()
}
```

File: packages/docusaurus-theme-classic/src/nprogress.css
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/nprogress.css:19]: We can now use the primary color instead of a hardcode color for the progress bar!

File: packages/docusaurus-theme-classic/src/nprogress.ts
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/nprogress.ts:27]: Technically, we can also stop the progress bar here, because at this time, the page is just about to re-render. But we can also use this to test `onRouteDidUpdate`.

File: packages/docusaurus/src/client/App.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/App.tsx:41]: No more hardcoded timeout constant in the core.

File: packages/docusaurus/src/client/PendingNavigation.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/PendingNavigation.tsx:70]: I don't know where's the best place to place this cleanup. Either in `render`, or in the callback of `setState` (where we previously had `this.stopProgressBar`)

**Review by Josh-Cena — COMMENTED** [review]
Introduced a new `onRouteDidUpdate` to retrieve the new DOM; `onRouteUpdate` will continue to fire as soon as route transition happens, like previously.

File: packages/docusaurus/src/client/PendingNavigation.tsx
Comment (Josh-Cena) [reply packages/docusaurus/src/client/PendingNavigation.tsx:70]: Actually... I'll move it to `setState`.

File: packages/docusaurus-types/src/index.d.ts
Comment (slorber) [inline packages/docusaurus-types/src/index.d.ts:608]: mark the PR as breaking change? although it shouldn't impact most users

File: packages/docusaurus/src/client/PendingNavigation.tsx
Comment (slorber) [inline packages/docusaurus/src/client/PendingNavigation.tsx:70]: not a really good practice to do such side effects in render 🤪 particularly with Concurrent React. 

I doubt here it causes any problem but I'd still rather move this to the preload callbacks instead

**Review by slorber — COMMENTED** [review]
👍 

Would move the cleanup callback call outside of render

Do we prefer `onRouteUpdate` + `onRouteDidUpdate` VS `onRouteWillUpdate` + `onRouteUpdate`? 🤷‍♂️ difficult choice

**Comment by slorber** [comment]
Also, don't we want ga/gtag plugin events to trigger only after React renders the next page? I think it would be better to wait, and not send any analytics event if the page transition never completes

File: packages/docusaurus-types/src/index.d.ts
Comment (Josh-Cena) [reply packages/docusaurus-types/src/index.d.ts:604]: Mmm, naming is hard. Between `onRouteUpdate`, `onRouteDidUpdate`, `onRouteWillUpdate`, I chose the former two because the third is slightly misleading: when the callback fires, the route update has already happened, it's just that the DOM hasn't re-painted yet.

File: packages/docusaurus/src/client/PendingNavigation.tsx
Comment (Josh-Cena) [reply packages/docusaurus/src/client/PendingNavigation.tsx:70]: Ah, sure

File: packages/docusaurus-theme-classic/src/nprogress.ts
Comment (Josh-Cena) [reply packages/docusaurus-theme-classic/src/nprogress.ts:27]: Let's keep it this way. Doesn't look too problematic

**Comment by Josh-Cena** [comment]
> Also, don't we want ga/gtag plugin events to trigger only after React renders the next page?

Migrated to `onRouteDidUpdate`. Impact should be minimal. There's a nuance: `onRouteUpdate` only fires when route changes, but `onRouteDidUpdate` would fire on initial render as well. Do you think that's a good thing, or it doesn't really matter?

File: packages/docusaurus-types/src/index.d.ts
Comment (slorber) [reply packages/docusaurus-types/src/index.d.ts:604]: I see, that makes sense 👍

**Comment by slorber** [comment]
> There's a nuance: onRouteUpdate only fires when route changes, but onRouteDidUpdate would fire on initial render as well. Do you think that's a good thing, or it doesn't really matter?

😭 I don't know 😅 

It looks like it does have an impact and lead to duplicate events being send on site first load 😓 

Before/prod:

CleanShot 2022-04-28 at 16 52 31@2x

After:

CleanShot 2022-04-28 at 16 53 09@2x

Maybe both lifecycles should never fire on the very first render when previousLocation is null?

I feel like having this difference between the 2 lifecycles will be confusing for users, and we should rather make the best choice now because this change later can also be painful for the plugin ecosystem. 


Note that client module authors can still add some inline JS that will be executed immediately after load:

```js
if (ExecutionEnvironment.canUseDOM) {
  console.log("on first load, document title is " + document.title);
}
```

Note that Gatsby seems to fire both lifecycles on first load: 
- https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/#onPreRouteUpdate
- https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/#onRouteUpdate
- https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/#onInitialClientRender

**Comment by slorber** [comment]
> We can fire it as frequently as we want, and ask client module authors to filter out unwanted triggers. For example, here we can add a simple `if (previousLocation === null)`. The only problem is I don't really know which is better off for the user: is it useful for analytics to know the first page the user is on, or should it only track navigations?

I guess it's more powerful to let the users figure out the filtering on their side in case duplicate events are fired on initial load.

Some analytics may send events automatically to their endpoint on js load, while others need to be triggered manually with a JS API. Giving the opportunity to trigger this API seems better to me 👍 

Now we should probably make both lifecycle consistent, and it's not very clear to me when/where `onRouteUpdate` should be fired in the app. It could be something like this?

```js
      if (previousLocation === null) {
        const cleanup = dispatchLifecycleAction('onRouteUpdate', {
          previousLocation,
          location,
        });
        cleanup();
      }
      
      dispatchLifecycleAction('onRouteDidUpdate', {previousLocation, location});
```

**Comment by Josh-Cena** [comment]
Okay, why not

**Comment by slorber** [comment]
seems good now 👍 

just need to filter first call on gtag/google-analytics plugins

File: packages/docusaurus-plugin-google-analytics/src/analytics.ts
Comment (Josh-Cena) [inline packages/docusaurus-plugin-google-analytics/src/analytics.ts:11]: This check actually isn't necessary before either...

**Comment by slorber** [comment]
tested locally, analytics work as before 👍 

great 🎉

**Comment by axe312ger** [comment]
Cool!

**Pull request merged**
